### PR TITLE
obs: add initial Project types

### PR DESCRIPTION
**Pull request by nitishfy**
Reference: [#2295](https://github.com/kubernetes/sig-release/issues/2295)
This PR is created with the purpose of adding the initial project structure for the OBS project meta file for golang library

### CODE REVIEW

File: obs/obs.go
Comment (nitishfy) [inline obs/obs.go:78]:
```suggestion
func main() {
```
This is just for testing purposes. We will remove it once this PR is ready to be merged.

**Review by nitishfy — COMMENTED** [review]
General question answered

File: .gitignore
Comment (xmudrii) [inline .gitignore:4]: Please add a new line here

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:10]: We don't need `yaml` tags, we can just use `json` tags with k-yaml library

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:13]: Title and Description are also optional

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:21]: What is Parent used for? I haven't seen it in our configs

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:24]: Let's have the API URL here as well

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:29]: Let's call this something like RepositoryPath to make it clear what is it used for

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:35]: Disable should be an empty struct instead

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:50]: Let's call this Person to match the OBS API

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:52]: Both should be `,attr`

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:57]: Let's have a plural name here for the consistency with the Project struct

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:58]: - There can be multiple release targets so this should be a slice
- Let's please use a full name, maybe ReleaseTargets

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:64]: Let's please use a full name, Repository

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:65]: Can you find what are valid values for this field and create a dedicated type, similar as for Role?

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:68]: Please make this an exported type and call it something like UserRole.

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:75]: Please make those variables exported.

Comment (xmudrii) [inline obs/obs.go:75]:
```suggestion
	roleBugOwner   role = "bugowner"
	roleMaintainer role = "maintainer"
	roleReviewer   role = "reviewer"
	roleDownloader role = "downloader"
	roleReader     role = "reader"
```

File: obs/obs.go
Comment (xmudrii) [reply obs/obs.go:78]: For follow-up: replace this with an unit test that can compare against a golden file

**Review by xmudrii — CHANGES_REQUESTED** [review]
Here's the initial review, I can do another review once these comments are addressed. Looks pretty solid so far!

File: obs/obs.go
Comment (nitishfy) [reply obs/obs.go:21]: This was just implemented for design on how a particular project will get to know who is the parent project of it. We can remove this field.

File: obs/obs.go
Comment (nitishfy) [reply obs/obs.go:57]: RepositoryName?

File: obs/obs.go
Comment (nitishfy) [reply obs/obs.go:65]: By now, I could spot only one value.

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:16]: These fields should have omitempty too

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:54]: I don't think we need omitempty here

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:61]: We don't need pointers here

File: obs/obs.go
Comment (xmudrii) [reply obs/obs.go:67]: I couldn't find anything in docs so let's leave it as it is

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:38]: Nit: move this either before or after structs that use it

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:78]: Let's call this type PersonRole and move type definitions close to the Person struct

**Review by xmudrii — CHANGES_REQUESTED** [review]
Looks better, some more nits from my side

File: obs/obs.go
Comment (nitishfy) [reply obs/obs.go:61]: but isn't this field as `omitempty`?

Comment (xmudrii) [reply obs/obs.go:61]: Yes, these fields are omitempty, but omitempty also works with slices out of the box without using pointers.

**Comment by nitishfy** [comment]
I think I can move forward with adding more features now. Let me know if you have any more feedback to address, I'll resolve them along the way.

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:54]: XML tag should be just `role` instead of `PersonRole`

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:14]: I changed my mind about these fields. Let's make this field non-pointer string and remove `omitempty`. We're actually going to get the same result.

File: obs/obs.go
Comment (xmudrii) [reply obs/obs.go:80]: Let's remove the `main` function and we can merge the PR once the other two comments are addressed.

**Review by xmudrii — CHANGES_REQUESTED** [review]
Some final nits for now. Perhaps we can merge this PR once these nits are addressed.

**Comment by xmudrii** [comment]
/assign @saschagrunert @cpanato 
PTAL at this PR :)

File: obs/obs.go
Comment (nitishfy) [reply obs/obs.go:14]: but how is this working internally then?

Comment (xmudrii) [reply obs/obs.go:14]: I'm not sure I fully understand your question. Can you elaborate please?

File: obs/obs.go
Comment (nitishfy) [reply obs/obs.go]: Done!

File: obs/obs.go
Comment (nitishfy) [reply obs/obs.go]: mhh, why is there need to remove the pointers and omitempty from here? (these fields are after all not mandatory, isn't it?)

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:40]: Linter says this should be `APIURL` so please change it :)

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:2]: Please change this to 2023.

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:91]: Let's prefix these variables with Person
```suggestion
	PersonRoleBugOwner   PersonRole = "bugowner"
	PersonRoleMaintainer PersonRole = "maintainer"
	PersonRoleReviewer   PersonRole = "reviewer"
	PersonRoleDownloader PersonRole = "downloader"
	PersonRoleReader     PersonRole = "reader"
```

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:86]: These definitions should be moved close to the Person struct.

File: obs/obs.go
Comment (xmudrii) [inline obs/obs.go:1]: Let's name this file `project.go` because it holds the Project-related structs. We'll use different files for different types.

File: obs/obs.go
Comment (xmudrii) [reply obs/obs.go]: Because working with string pointers (or with any other pointer to a primitive type) is a little bit complicated and doesn't provide good user experience. When defining the struct, you have to do something like this:

```go
str := "My Project Name"
p := &Project{
    URL: &str,
}
```

Instead of:

```go
p := &Project{
    URL: "My Project Name",
}
```

When you have multiple strings, like in our case, it gets even worse.

Right now, if you completely omit the URL when defining a project, you'll get XML like this one:

```xml
<project>
    ...
</project>
```

If you remove pointers and `omitempty`, you'll get XML like this one:

```xml
<project>
    <url></url>
</project>
```

These two XMLs are practically the same. OBS is even going to remove the empty tags automatically. However, the user experience of using the library is much better in the second case (when not using pointers to primitive types).

**Comment by nitishfy** [comment]
**Update:** PR is ready for the next set of review and probably for merge as well.
cc: @saschagrunert @cpanato

File: obs/project.go
Comment (xmudrii) [inline obs/project.go:80]: Nit: JSON tag should be `releaseTarget`:
```suggestion
	ReleaseTargets []ReleaseTarget  `json:"releaseTarget,omitempty" xml:"releasetarget,omitempty"`
```

File: obs/project.go
Comment (xmudrii) [inline obs/project.go:79]: Nit: JSON tag should be `architectures`:
```suggestion
	Architectures  []string         `json:"architectures" xml:"arch"`
```

**Review by xmudrii — APPROVED** [review]
Some nits, but let's address these nits in a follow-up.
/lgtm
/approve

**Pull request merged**
